src/java/org/apache/cassandra/index/sai/plan/Expression.java

```diff
@@ -106,244 +110,344 @@ public Expression(IndexContext indexContext)
-                    lower = new Bound(value, validator, lowerInclusive);
-                else
-                    upper = new Bound(value, validator, upperInclusive);
-                break;
+        private final StorageAttachedIndex index;
 
-            case GTE:
-                if (context.getDefinition().isReversedType())
-                {
-                    this.upperInclusive = true;
-                    upperInclusive = true;
-                }
-                else
-                {
-                    this.lowerInclusive = true;
-                    lowerInclusive = true;
-                }
-            case GT:
-                operator = IndexOperator.RANGE;
-                if (context.getDefinition().isReversedType())
-                    upper = new Bound(value, validator, upperInclusive);
-                else
-                    lower = new Bound(value, validator, lowerInclusive);
-                break;
-            case ANN:
-                operator = IndexOperator.ANN;
-                lower = new Bound(value, validator, true);
-                upper = lower;
-                break;
+        public IndexedExpression(StorageAttachedIndex index)
+        {
+            super(index.termType());
+            this.index = index;
+        }
+
+        @Override
+        public boolean isNotIndexed()
+        {
+            return false;
         }
 
-        assert operator != null;
+        @Override
+        public StorageAttachedIndex getIndex()
+        {
+            return index;
+        }
 
-        return this;
+        @Override
+        public AbstractAnalyzer getAnalyzer()
+        {
+            return index.analyzer();
+        }
     }
 
-    /**
-     * Used in post-filtering to determine is an indexed value matches the expression
-     */
-    public boolean isSatisfiedBy(ByteBuffer columnValue)
+    class UnindexedExpression extends AbstractExpression
     {
-        // If the expression represents an ANN ordering then we return true because the actual result
-        // is approximate and will rarely / never match the expression value
-        if (validator.isVector())
+        private UnindexedExpression(IndexTermType indexTermType)
+        {
+            super(indexTermType);
+        }
+
+        @Override
+        public boolean isNotIndexed()
+        {
             return true;
+        }
 
-        if (!TypeUtil.isValid(columnValue, validator))
+        @Override
+        public StorageAttachedIndex getIndex()
         {
-            logger.error(context.logMessage("Value is not valid for indexed column {} with {}"), context.getColumnName(), validator);
-            return false;
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public AbstractAnalyzer getAnalyzer()
+        {
+            return new NoOpAnalyzer();
         }
+    }
+
+    abstract class AbstractExpression implements Expression
```

Comment:
I'm not sure there is a benefit in having a separate interface and abstract class for this.

Reply:
Agreed, I have made `Expression` an `abstract class`.